File: day2/day2.py
import re
import sys

print('Loading file:', str(sys.argv[1]))

# Read integers from the file
lines = []
with open(str(sys.argv[1])) as f:
    lines = f.read().splitlines()

items = []
for line in lines:
    # group 1: min
    # group 2: max
    # group 3: character
    # group 4: password
    m = re.match(r'(\d+)-(\d+)\s(.*):\s(.*)', line)
    items.append((int(m.group(1)), int(m.group(2)), m.group(3), m.group(4)))

pass_count = 0
for item in items:
    min_val = item[0]
    max_val = item[1]
    char = item[2]
    password = item[3]

    first = False
    try:
        first = password[min_val-1] == char
    except:
        pass

    second = False
    try:
        second = password[max_val-1] == char
    except:
        pass

    if first ^ second:
        pass_count += 1

print(pass_count)

File: day2/day2.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	filename := os.Args[1]

	input, err := ioutil.ReadFile(filename)
	if err != nil {
		fmt.Println("File reading error", err)
		return
	}

	text := string(input)
	lines := strings.Split(text, "\n")

	// Remove empty lines
	temp := lines[:0]
	for _, x := range lines {
		if len(x) > 0 {
			temp = append(temp, x)
		}
	}
	lines = temp

	re, _ := regexp.Compile(`(\d+)-(\d+)\s(.*):\s(.*)`)

	// Convert strings to integers
	var validCount = 0
	for _, line := range lines {
		result := re.FindStringSubmatch(line)

		min, _ := strconv.Atoi(result[1])
		max, _ := strconv.Atoi(result[2])
		character := result[3]
		password := result[4]

		f := string(password[min-1])
		s := string(password[max-1])

		if (f == character) != (s == character) {
			validCount++
		}
	}

	fmt.Println(validCount)
}
